### ISSUE TITLE
Finding SoftMax activations

### ISSUE THREAD

**Issue by sebastian-lapuschkin** [albermax/innvestigate#84]
One requirement for models to use in this toolbox is that they do not have any softmax-activated outputs.
Since keras allows dedicated SoftMax layers as well as SoftMax activations to be attachted to e.g. dense layers, the warning messages upon creating an analyzer object can be a bit unspecific.

How about the implementation of marginally more informative messages regarding the type of softmax activation found (is it a layer? is it part of a layer) and where it occured (print  the layer index or name).

This should help debugging.
Thanks!

**Comment by enryH** [comment]
Does this then also include sigmoid activations applied to the last dense layer? 

Could you maybe give a hint why this is not allowed?

**Comment by sebastian-lapuschkin** [comment]
Hi @enryH ,

sigmoid activations are (as far as I know) not covered yet by this  restriction, but should be for the same reason as the softmax output.

The (most important) reason is::
SoftMax applies a non-linear scaling to the output which outputs probability scores. This might help during training and provides some nice looking predictions during inference, but does not change the ranking order of the predicted classes (assuming a classification task).

The exponential scaling however, might  obfuscate the *true* confidence of the model's *individual* class predictions. Assume a predictor output for two classes from a model trained to predict the presence of a class if the respective `y>0`:

```
logit_1 = [1,1]
->
softmax_1 = [0.5, 0.5]

logit_2 = [1,0]
->
softmax_2 = [0.73.., 0.26..]

logit_3 = [0,0]
->
softmax_3 = [0.5, 0.5]

logit_4 = [-1,-2]
->
softmax_4 = [0.73.., 0.26..]
```

The problems we see here is that the softmax does not allow an independent assessment of model outputs: Giving trash inputs to a model results in the network picking one with a (very) high probability, although all last layer logits indicate no class membership at all.
This behaviour makes the prediction hard to analyze meaningfully.

**Comment by enryH** [comment]
Thanks for the hint on softmax! Great example to understand softmax behaviour:)

Regarding sigmoid: Since it lacks the "logits specific" normalization as softmax, the rescaling should not be a too big problem then, I assume? (so maybe no warning for sigmoid output activations?)

**Comment by sebastian-lapuschkin** [comment]
regarding sigmoids:
I can only argue in terms of LRP behaviour right now.

The models I used until now usually had some softmax output activation or just returned the (last-step-linear) logits.
In general, adding an activation to the terminal neurons of a network should not be a problem.
However, I assume that this should only work (reliably) with monotonously increasing activation fxns (which most (all?) should be), which fulfill the condition of f(0) = 0.

the reason is: given some negative logit activations, put into the logistic sigmoid (maps f(0) = 0.5, projects in general into the $R^+$) returns positive output activations.
since the sigmoid is a component-wise functions, the LRP backward pass would always backpropagate positive relevance values to the logit output neurons.

now assume your logit of interest was negatively valued (which, after the application of the sigmoid, becomes a positive value, thus R_j is positive).
Applying LRP to the logit directly would yield meaningful heatmaps (negative scores for inputs making the output small/negative, positive ouput making them large/positive).
The sigmoid output inverts the sign (and applies some scaling) of the model output you analyze, thus would sign-invert every hidden neuron relevance and input relevance compared to the logit-based analysis.


having sigmoids or really any kind of component-wise activation function within the network (hidden units) is not a problem however since
1) the just pass along the relevance and
2) the non-linearities of the net manifest in the inputs to linear components of the net (dense, conv, pooling layers), jet LRP operates step-wise linear. since the inputs to each decomposition step are non-linearly activated, LRP computes over the whole model non-linear decompositions.


Since DTD basically is LRP with  more strict assumptions I would assume similarly, that a sigmoid-output as part of the analysis would cause problems with negative logits feeding the sigmoid.

There is, however, no reason you should not be able to use the sigmoid layer for prediction and remove it for analysis, since usually, small (negative) logit values = "no" decision and large (positive) logit values = "yes".

Best,

**Comment by albermax** [comment]
Fixed.
